feat(CategoryTheory/Topos): define elementary topos

fweth opened the pull request:
This commit begins the formalization of the notion of an *elementary topos*, following the definition in \[Mac Lane & Moerdijk, *Sheaves in Geometry and Logic* (1992), Ch. IV, Section 1]. It introduces the file `Mathlib/CategoryTheory/Topos/Basic.lean`, which currently includes:

* The definition of an elementary topos.
* Proof that the power object map is a functor
* Theorem about pullbacks of characterstic maps

Subobject classifiers, which are used in the definition, have already been defined in `CategoryTheory/Topos/Classifier.lean`

Work in progress: further formalization of the section is planned.

Questions:

* Should the definition of power objects be separated into its own file/module, perhaps under `CategoryTheory/Topos/PowerObject.lean`?
* Is the notation `P` / `P_morph` / `P_functor` for the power object functor on objects / morphisms / total acceptable?
* Should the comments and docstrings rather use the notation from the book, where `g ∘ f` denotes arrow composition from righ to left, or the Lean variant `f ≫ g`?

Klaus Gy klausgy@gmail.com

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

fweth committed:
...

fweth committed:
...

fweth committed:
started functor definition

fweth committed:
...

fweth committed:
pullback_of_char

fweth committed:
new notation

fweth committed:
layout

fweth committed:
minor docs change

fweth committed:
definition doesnt need to be noncomputable

fweth committed:
chore: lake exe mk_all

fweth committed:
added docstring

fweth committed:
copyright added

fweth committed:
more elegant proof using simpa

fweth committed:
merged master

joelriou commented on Mathlib/CategoryTheory/Topos/Basic.lean:
```diff
@@ -0,0 +1,122 @@
+  hc : Classifier ℰ
+  /-- Power objects, will become a functor `P : ℰᵒᵖ ⥤ ℰ` later . -/
+
+  P (B : ℰ) : ℰ
+  /-- The element relation. -/
+  ε_ (B : ℰ) : B ⨯ (P B) ⟶ hc.Ω
+  /-- The P-transpose of a morphism `f : B × A ⟶ Ω`. See equation (6) of MM92. -/
+  unhat {A B : ℰ} (f : B ⨯ A ⟶ hc.Ω) : (A ⟶ P B)
+  /-- Characteristic equation: any `f : B × A ⟶ Ω` is equal to `ε_B ∘ (𝟙 ⨯ g)`
+      where `g` is the P-transpose of `f`. -/
+  comm {A B : ℰ} (f : B ⨯ A ⟶ hc.Ω) :
+    f = (𝟙⨯ (unhat f)) ≫ ε_ B
+  /-- Uniqueness: the P-transpose `g : A ⟶ P B` is uniquely determined by `f`. -/
+  uniq {A B : ℰ} (f : B ⨯ A ⟶ hc.Ω) (g : A ⟶ P B)
+    (_ : f = (𝟙⨯ g) ≫ ε_ B) : g = (unhat f)
```
This seems a little bit ad hoc. I think that this should be first defined in the context of monoidal categories. If `B` and `C` are two objects of a monoidal category, the question is whether the functor which sends `A` to `B ⊗ A ⟶ C` is representable by an object `P`. (This is related, but not exactly what we have in `CategoryTheory.Closed.Monoidal`. In case `B` is "closed" (or if the category if monoidal closed), then `P` would be `(ihom B).obj C`. Here, we only require that the "right adjoint" of `B ⊗ -` is "defined" at `C`) 
The power object seems to be the particular case `C := hc.Ω`.

In the general context of monoidal categories, we should have a condition (not necessarily a `Prop`) which depends on `B` and `C` which says that a certain functor satisfies `RepresentableBy`. That this holds for all `B` for a specific `C` is also interesting condition, under which "we can define the flipped internal hom (bi)functor evaluated at `C`".

Arguably, instead of the cartesian product given by the limits API, I believe we could use the `CartesianMonoidalCategory` API.

fweth replied:
Is this better?

```
open CategoryTheory Category Functor Limits MonoidalCategory Opposite

variable {ℰ : Type u} [Category.{v} ℰ] [CartesianMonoidalCategory ℰ]

def WhiskeredHom (B C : ℰ) : ℰᵒᵖ ⥤ Type v :=
  ⟨ ⟨ fun A ↦ B ⊗ unop A ⟶ C, fun f g ↦ (B ◁ unop f) ≫ g ⟩,
    fun A ↦ by
      ext g
      have : unop (𝟙 A) = 𝟙 (unop A) := by rfl
      simp[this],
    fun f f' ↦ by
      ext g
      have : B ◁ unop (f ≫ f') = B ◁ unop f' ≫ B ◁ unop f := by aesop_cat
      simp[this] ⟩

def IsPowerObjectOf (hc : Classifier ℰ (𝟙_ ℰ)) (B P : ℰ) :=
  RepresentableBy (WhiskeredHom B P) (hc.Ω)

variable [HasPullbacks ℰ]

structure ElementaryTopos where
  hc : Classifier ℰ (𝟙_ ℰ)
  P (B : ℰ) : ℰ
  power_object (B : ℰ) : IsPowerObjectOf hc B (P B)
```

fweth committed:
Merge remote-tracking branch 'upstream/master' into topos-definition

fweth committed:
Merge remote-tracking branch 'upstream/master' into topos-definition

fweth committed:
definition as representable functor

fweth committed:
removed commented out

fweth committed:
added doc strings

fweth committed:
deleted temp file

fweth committed:
added whisker notation

fweth committed:
localized whisker definition

fweth committed:
pairing notation

fweth committed:
notation

fweth committed:
fixed isPowerobjectOf

fweth committed:
docstring

fweth committed:
class instead of structure

fweth committed:
some more definitions

fweth commented:
Removing the draft label for now, although there are still some lemmas I added which need to be dualized. Also there is some stuff I proved in my original, 'ad hoc' classifier structure, which I need to port. Also plan to add an alternative constructir, more similar to how an elementary topos is defined in the SiGaL book (or what I had in my 'ad hoc' construction earlier).

-awaiting-author

fweth committed:
docstring

fweth commented:
Also I noticed this peculiar thing which I cannot explain:

```
lemma isPullback_equalizer_prod' [HasBinaryProduct Y Y] (e : Fork f g) (h : IsLimit e) :
    IsPullback e.ι (e.ι ≫ f) (prod.lift f g) (diag _) := by
  let toFork (s : PullbackCone (prod.lift f g) (diag Y)) :=
    Fork.ofι s.fst (prod.lift_eq_diag_of_comp_eq (s.condition))
  refine ⟨⟨by ext <;> simp [e.condition]⟩, ⟨PullbackCone.IsLimit.mk _ ?_ ?_ ?_ ?_⟩⟩
  · exact fun s ↦ h.lift (toFork s)
  · exact fun s ↦ Fork.IsLimit.lift_ι h
  · exact fun s ↦ by calc
      _ = s.fst ≫ (prod.lift f g) ≫ prod.fst := by simp; rfl
      _ = s.snd := by rw[reassoc_of% s.condition]; simp
  · exact fun _ _ hm _ ↦ Fork.IsLimit.hom_ext h (Eq.symm (Fork.IsLimit.lift_ι h) ▸ hm)
  ```
  
  this proof works fine but when I inline `toFork` instead, that means
  
  ```
    · exact fun s ↦ h.lift (Fork.ofι s.fst (prod.lift_eq_diag_of_comp_eq (s.condition)))
```

I get a maximum recursion depth error in `simp` later. Why would that be the case?

fweth committed:
Merge remote-tracking branch 'upstream/master' into topos-definition

fweth committed:
typo in docstring

fweth committed:
removed definition

fweth committed:
cleaner proof, removed simp lemmas but there is another pr for it

fweth committed:
Merge remote-tracking branch 'upstream/master' into topos-definition

fweth committed:
comm lemma added

fweth committed:
changed direction of comm lemma

fweth committed:
fix

fweth committed:
defined p functor

fweth commented:
Do we want a functor `P` in the `ElementaryTopos` structure or just an object map `P`? The object map makes a lot of notation easier, like `P B` instead of `P.obj (op B)` and we can derive functoriality as a lemma. What is preferred?

fweth committed:
modularized api, pointwise powerobject constructor

fweth committed:
docstring

fweth committed:
streamline

fweth committed:
added abbrevs

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

fweth committed:
structural changes

fweth committed:
merged upstream

fweth committed:
fixed structure

fweth committed:
more defs

fweth committed:
docstring

fweth committed:
char_of_pullback + cleanup

fweth commented:
If I'm not mistaken, there is a lot of stuff which needs to be duplicated for cartesian monoidal categories. Is that the long term goal of mathlib, to duplicate all the existing theory about binary products also for cartesian monoidal categories?

fweth committed:
created categorytheory.closed.powerobjects

fweth commented:
I've now moved the power object definitions into `Mathlib/CategoryTheory/Closed/PowerObjects.lean`, is that what you had in mind?

> In principle, we should state definitions/lemmas using general (co)fans/(co)cocones, and deduce versions for the (chosen) binary products.

I agree, but doing so touches a lot of the foundational product and equalizer APIs. I needed it mostly for my `char_of_pullback` lemma, which I now moved into a private branch, and will bring into a new PR after completing more groundwork, if you agree.

-awaiting-author

joelriou commented on Mathlib/CategoryTheory/Closed/PowerObjects.lean:
```diff
@@ -0,0 +1,100 @@
+
+variable {ℰ : Type u} [Category.{v} ℰ] [CartesianMonoidalCategory ℰ]
+
+/-- The covariant functor `B ⊗ [] ⟶ C` from `ℰᵒᵖ` to `Type`. -/
+def WhiskeredHom (B C : ℰ) : ℰᵒᵖ ⥤ Type v :=
+  ⟨ ⟨ fun A ↦ B ⊗ unop A ⟶ C, fun f g ↦ (B ◁ unop f) ≫ g ⟩,
+    fun A ↦ by
+      have : unop (𝟙 A) = 𝟙 (unop A) := by rfl
+      ext; simp[this],
+    fun f f' ↦ by
+      have : B ◁ unop (f ≫ f') = B ◁ unop f' ≫ B ◁ unop f := by aesop_cat
+      ext; simp[this] ⟩
+
+/-- `P` is a power object of `B` if it represents the functor `WhiskeredHom B hc.Ω`. -/
+def IsPowerObjectOf (sc : Classifier ℰ (𝟙_ ℰ)) (B P : ℰ) :=
+  (WhiskeredHom B sc.Ω).RepresentableBy P
```
This has to be more general. There should be no reference to `Classifier`, and I believe it should apply to any monoidal category.

fweth replied:
The alternative, as far as I can see, wolud be this:

```lean
variable {ℰ : Type u} [Category.{v} ℰ] [HasPullbacks ℰ] [CartesianMonoidalCategory ℰ]

noncomputable def WhiskeredSubobject (B : ℰ) : ℰᵒᵖ ⥤ Type (max u v) where
  obj A := Subobject (B ⊗ unop A)
  map f := (Subobject.pullback (B ◁ unop f)).obj
  map_id A := by
    ext1 x
    simp [show unop (𝟙 A) = 𝟙 (unop _) from rfl, Subobject.pullback_id]
  map_comp f f' := by
    ext1 x
    simp [show unop (f ≫ f') = unop f' ≫ unop f from rfl, Subobject.pullback_comp]
```

Is this better, should I continue with this?

fweth committed:
dualized lemmas

fweth committed:
fixed simp lemma

fweth committed:
rewrite of power objects

fweth committed:
merge upstream

fweth committed:
power objects via left representability

fweth committed:
removed some imports

fweth committed:
resolved conflicts with upstream

fweth committed:
update lean version and fix compose lemma

fweth committed:
Merge remote-tracking branch 'upstream/master' into topos-definition

fweth committed:
fixed subobject/basic.lean

fweth committed:
remvoe deleted file from index

fweth committed:
docstring

fweth committed:
docstring

fweth committed:
long file option for commsq.lean

fweth committed:
run exe shake --fix

fweth committed:
removed import

fweth committed:
removed epsilon

fweth commented:
Hi @joelriou,

sorry for the longer silence. I think I found a nice way now to introduce power objects: First, I show that for any functor `F : ℰᵒᵖ × ℰᵒᵖ ⥤ Type` and `P : ℰ → ℰ` such that `P B` represents `F(B, *)` we can turn `P` into a functor `ℰᵒᵖ ⥤ ℰ`. Then, in the case of power objects, we chose `F(B, C) := Subobject (B ⊗ C)` and immediately get functoriality.

-awaiting-author

fweth committed:
removed paranthesis

fweth committed:
better simp

fweth committed:
better docstring

fweth committed:
delete diff.txt

fweth committed:
generalize isoIsPullback_congr lemma

fweth committed:
docstring

fweth committed:
Merge remote-tracking branch 'upstream/master' into topos-definition

fweth committed:
rw

fweth committed:
rename iota to iso

fweth commented:
I'm currently trying to rewrite from `F : ℰᵒᵖ × ℰᵒᵖ ⥤ Type` to `F : (ℰ × ℰ)ᵒᵖ ⥤ Type`, I think that's cleaner, than we can later use the already defined subobject functor on `ℰ × ℰ`.

fweth committed:
merge upstream

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

fweth committed:
removed xclip

fweth committed:
merge
